# Complete lecture notebook — 2

Read the explanation, predict the result, run the next cell, and interpret its output before continuing. Use **Restart Kernel and Run All** to check that the work does not depend on an old session. The complete chapter narrative, mathematical derivations and physical assumptions are retained around the cells.

## Supplied notebook setup
Select **Julia 1.12**. Start this notebook from its chapter directory. This supplied setup locates the recorded course environment and creates a separate folder for exported figures; it installs no packages. Setup commands are infrastructure, not new assessed language material.

In [1]:
import Pkg
course_root = abspath("../..")
if !isfile(joinpath(course_root, "Project.toml"))
    error("Start this notebook from its week folder, e.g. week-1/notebooks")
end
Pkg.activate(course_root; io=devnull)
output_dir = get(ENV, "COURSE_NOTEBOOK_OUTPUT", joinpath(pwd(), "generated", "lecture-en"))
mkpath(output_dir)
println("Course environment active; figure output directory ready.")

Course environment active; figure output directory ready.


# Week 2: Values, Decisions, and a Block on an Incline

## Lecture: From Variables to an Algorithm

In Week 1 we gave names to measurements and calculated new values from them. Before adding decisions, we will slow down and make that basic idea explicit: what a variable is, what assignment does, and how values change as Julia executes one line after another.

### A variable gives a value a name

A variable is a name that refers to a value. A useful name lets us discuss the meaning of a number instead of repeatedly writing the number itself:

Assignment has the general form `name = expression`. Julia calculates the expression on the right and makes the name on the left refer to the result. The smallest example is `x = 2`. A richer scientific calculation uses several meaningful names:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [2]:
mass = 2.0        # kg
speed = 3.0       # m/s
momentum = mass * speed   # kg m/s

println(momentum)   # 6.0

6.0


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
6.0
```

The single equal sign is the *assignment operator*. Julia first evaluates the expression on the right of `=`; it then makes the name on the left refer to the resulting value. Thus the third line first multiplies `mass` by `speed`, then assigns the result to `momentum`.

#### Momentum of a moving mass

An object's momentum measures how much motion it carries: a heavier object and a faster object each carry more. For a point mass moving well below the speed of light, $\mathbf{p}=m\mathbf{v}$, with magnitude $p=mv$ along the direction of motion. Mass $m$ is in kilograms, speed $v$ in metres per second, and momentum in kilogram-metres per second (kg m/s).

We can inspect a variable simply by writing its name in a notebook cell, or print it as part of a message:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

These are separate interactive expressions. We place them in separate cells so each result is visible, rather than displaying only the last one.

In [3]:
momentum

6.0

In [4]:
println("momentum = $(momentum) kg m/s")

momentum = 6.0 kg m/s


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

In a notebook the first line displays `6.0`; the printed report is:

Reference output in the written notes (compare with the actual cell output):

```text
momentum = 6.0 kg m/s
```

### Julia follows the program in order

An assignment performs a calculation at that moment. It does not create a permanent algebraic relationship between variables:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [5]:
x = 4
y = x + 1       # y is now 5
x = 10          # x changes, but y is still 5

10

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

When Julia executes `y = x + 1`, it uses the current value of `x`. Changing `x` later does not automatically recalculate `y`; we would need to execute `y = x + 1` again. This line-by-line order is especially important in notebooks, where cells can be run more than once or out of order.

### A variable can be assigned a new value

The same name can refer to a new value later in the program:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [6]:
temperature = 20.0   # C
temperature = temperature + 0.5

println(temperature)     # 20.5

20.5


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
20.5
```

The second line is not an algebraic equation. Julia reads the old value on the right, adds `0.5`, and assigns the new result back to the same name. Loops will use this pattern to advance time, increase a counter, or update a simulated position.

#### Names should communicate meaning

**Names**: Variable names are case-sensitive, cannot contain spaces, and should be used consistently. You can use unicode, and in particular greek symbols (e.g., type `\phi<Tab>` to get $\phi$). Prefer meaningful names to unexplained names such as `a`, `b`, and `c`. **Units**: Julia generally stores just ordinary numbers without their units. However in physics, it is advisable to keep track of them. You can add unit suffixes to the names, e.g., `m_kg`, or alternatively, as we do here, record the unit in a short trailing comment where the value is defined, for example `time = 0.35 # s`, `height = 1.20 # m`, and `g = 9.81 # m/s^2`; This is not compulsory but very useful to avoid unit errors. In bonus assignment 2 (i.e., not part of the official course material), we will see that we can use a package `Unitful` that adds units to variables, allowing us to keep track them. Units should likewise be stated in function docstrings or interface documentation, and clarified at call sites when otherwise ambiguous.

### Values have types

Variables can name different kinds of values. We already used an integer count, decimal measurements, and text labels in Week 1. This week we add Boolean comparisons and make four distinctions explicit because a program that reads input, chooses a branch, and decides when to stop must know what kind of value it is handling.

A *type* is a category of values that determines how Julia stores them and which operations make sense. For example, `5`, `0.35`, `"height"`, and `true` have different formats and represent an integer, a floating-point number, text, and a true/false value.

### Four simple value types

#### The four types we need now

  Julia type   Example                  Typical scientific role
  ------------ ------------------------ ----------------------------------------
  `Int`        `n = 5`                  counts and loop indices
  `Float64`    `t = 0.35 # s`           measurements and calculated quantities
  `String`     `label = "height [m]"`   input text, labels, and messages
  `Bool`       `ready = true`           true/false answers to conditions

The function `typeof(value)` lets us inspect a value's type. We use it for diagnosis, not as a replacement for understanding what the value means.

On the usual 64-bit course computers, `typeof(5)` prints `Int64`. Julia's name `Int` means the standard integer size for the computer and is an alias for `Int64` on these machines.

These types connect directly to Week 1:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [7]:
current = [0.10, 0.20, 0.30, 0.40, 0.50]   # A
n = length(current)                 # Int
I_bar = sum(current) / n            # Float64, A
axis_label = "current I [A]"    # String
analysis_ready = true           # Bool

println(typeof(n))
println(typeof(I_bar))
println(typeof(axis_label))
println(typeof(analysis_ready))

Int64


Float64
String
Bool


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

An integer and a floating-point number can participate in one calculation. Julia converts the integer as needed:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

These are separate interactive expressions. We place them in separate cells so each result is visible, rather than displaying only the last one.

In [8]:
2 * 0.35       # 0.7, a Float64

0.7

In [9]:
5 / 2          # 2.5, a Float64

2.5

In [10]:
Float64(5)     # 5.0

5.0

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

For now, these practical rules are sufficient. We do not need type annotations, abstract types, or the details of floating-point storage.

#### A type is not a physical unit

In `t = 0.35 # s`, the Julia type is `Float64`; the physical unit is seconds. The unit lives in the comment and in the relevant interface documentation, so a different `Float64` could represent metres or kilograms. Type checking and dimensional analysis answer different questions, and we need both.

### Comparisons produce Boolean values

Programs need yes/no questions in order to validate data and choose what to do next. A comparison has the form `left_value operator right_value` and produces a `Bool`: either `true` or `false`. Start with direct questions whose answers are obvious:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

These are separate interactive expressions. We place them in separate cells so each result is visible, rather than displaying only the last one.

In [11]:
2 < 3          # true

true

In [12]:
2 == 3         # false

false

In [13]:
5 != 4         # true

true

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

The main comparison operators are `<`, `<=`, `>`, `>=`, `==` (equal), and `!=` (not equal). The single equal sign `=` assigns a value; the double equal sign `==` asks a question.

We can combine Boolean values when one comparison is not enough. The operators `&&`, `||`, and `!` mean and, or, and not:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

These are separate interactive expressions. We place them in separate cells so each result is visible, rather than displaying only the last one.

In [14]:
true && false      # false: both must be true

false

In [15]:
true || false      # true: at least one is true

true

In [16]:
!true              # false

false

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

The same formats can express a richer physical validity rule:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [17]:
t = 0.35   # s
z0 = 1.20   # m
nonnegative_time = t >= 0
valid_height = z0 >= 0
valid_input = nonnegative_time && valid_height

true

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

### A first decision with if

Calculating a Boolean answer is useful only if the program can act on it. An `if` statement is a conditional: it runs one block when its condition is `true`. An optional `else` block runs when the condition is `false`. The general format is:

### Syntax template
This shows the general form; `condition` is a placeholder. Read it here, then run the concrete example that follows.

In [ ]:
if condition
    # commands when the condition is true
else
    # commands when the condition is false
end

Every conditional ends with `end`. If nothing should happen in the false case, we omit the `else` block. Here is the smallest two-way example:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [18]:
temperature = 25.0   # C
if temperature > 30
    println("hot")
else
    println("not hot")
end

not hot


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
not hot
```

Sometimes continuing would make a calculation meaningless. Julia provides a deliberate stop with this format:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

### A deliberate error, observed safely
The chapter deliberately stops a calculation here. The supplied `try`/`catch` wrapper prints that expected error and allows Run All to continue; unexpected errors are rethrown. We focus on the failing instruction, not on learning the wrapper yet.

In [19]:
try
    throw(ArgumentError("message"))
catch err
    err isa ArgumentError || rethrow()
    showerror(stdout, err)
    println()
end

ArgumentError: message


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

It stops execution and reports why. We will use it after a failed conversion and after physically invalid input.

## Piecewise Kinematics as a Decision

#### Signed acceleration versus positive gravity magnitude

Height is positive upward, so gravitational acceleration is $$g=-9.81\ \mathrm{m/s^2}.$$ For initial height $z_0$ and velocity $v_0$, $$z_{\mathrm{free}}(t)=z_0+v_0t+\frac{1}{2}gt^2.$$ The three terms have units of metres. This dimensional check is independent of whether their code values are `Float64`.

Our no-bounce model does not allow the ball to continue below the ground: $$z(t)=\begin{cases}
z_{\mathrm{free}}(t),&z_{\mathrm{free}}(t)>0,\\
0,&z_{\mathrm{free}}(t)\leq0.
\end{cases}$$

The piecewise definition translates directly to a conditional:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [20]:
t = 0.35   # s
z0 = 1.20   # m
v0 = 0.0    # m/s
g = -9.81   # m/s^2

if t < 0 || z0 < 0
    throw(ArgumentError("time and initial height must be non-negative"))
end

z_free = z0 + v0 * t + 0.5 * g * t^2   # m

if z_free > 0
    z = z_free
    state = "in air"
else
    z = 0.0
    state = "on the ground"
end

println("at t = $(t) s: z = $(round(z, digits=3)) m ($(state))")

at t = 0.35 s: z = 0.599 m (in air)


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
at t = 0.35 s: z = 0.599 m (in air)
```

At $t=0.35\ \mathrm{s}$ this program reports approximately $z=0.599\ \mathrm{m}$. The result passes three checks: time is in the model domain, the height is below its initial value, and every term in the position equation has units of metres.

### Several branches

Use `elseif` when more than two cases are mutually exclusive and should be tested in order. The general format is `if`, followed by any number of `elseif` branches, an optional `else`, and one `end`. Only the first true branch runs. A small sign-classification example shows the structure:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [21]:
value = -2
if value < 0
    description = "negative"
elseif value == 0
    description = "zero"
else
    description = "positive"
end

"negative"

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

The same structure gives a more useful height classification:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [22]:
if z == 0
    state = "on the ground"
elseif z < 0.20
    state = "close to the ground"
else
    state = "in air"
end

"in air"

**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Nested decisions are sometimes necessary, but a flat `if`/`elseif`/`else` chain is easier to inspect when it expresses the same logic.

#### Three error categories

- **Syntax error:** Julia cannot parse the program, for example because an `end` or closing parenthesis is missing.

- **Runtime error:** the program starts and then reaches an invalid operation, for example raising a `String` to a power.

- **Logical error:** the program runs but implements the wrong algorithm or physics, for example using positive $g$ while height is positive upward, or failing to advance time in a `while` loop.

Read the exact error and line number, isolate the smallest failing block, inspect values and types, check dimensions and signs, and test cases whose answers are already known.

We now make programs choose among physically distinct cases. Before a calculation can be trusted, we must know what kind of value we hold, convert deliberately, and state the domain in which a model applies.

## Values have types

Julia values carry types. Integers, floating-point numbers, text, and truth values serve different purposes.

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [23]:
println(3, "  ->  ", typeof(3))
println(3.0, "  ->  ", typeof(3.0))
println("3", "  ->  ", typeof("3"))
println(3 < 4, "  ->  ", typeof(3 < 4))

3  ->  Int64
3.0  ->  Float64
3  ->  

String
true  ->  Bool


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
3  ->  Int64
3.0  ->  Float64
3  ->  String
true  ->  Bool
```

The exact integer type can depend on the computer; the important distinction is `Int`, floating-point, `String`, and `Bool`.

An integer and a floating-point number can mix in ordinary arithmetic; Julia promotes the result where needed. Text is not silently a number. We choose conversions explicitly in both directions.

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [24]:
println(2 + 0.5)
text = "12.5"
mass = parse(Float64, text)   # kg
println(mass + 0.5)
println(string(mass, " kg"))
println(Int(7.0))

2.5
13.0
12.5 kg
7


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
2.5
13.0
12.5 kg
7
```

`parse(Float64, "twelve")` throws an error because the text is not a floating-point numeral. `Int(7.2)` also throws an error: converting to `Int` is allowed only when the value is exactly integral. This protects us from quietly discarding information.

## Booleans and gradual decisions

Comparisons produce Boolean values. We first ask one question, then choose between two cases, then add a third case with `elseif`.

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [25]:
temperature = -2.0   # C
println(temperature < 0)
if temperature < 0
    println("below freezing")
else
    println("not below freezing")
end

true


below freezing


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
true
below freezing
```

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [26]:
if temperature < 0
    state = "ice"
elseif temperature == 0
    state = "melting-point"
else
    state = "liquid"
end
println(state)

ice


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
ice
```

`ice`. Exactly one arm of an `if`/`elseif`/`else` chain runs. `=` assigns; `==` compares.

We combine conditions with `&&` (both), `||` (at least one), and `!` (not). A validity check often asks whether a value is a finite number. `isfinite(value)` returns `true` for ordinary finite numbers and `false` for `NaN`, `Inf`, or `-Inf`, which are not usable measurements. We can combine that check with a physical condition:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [27]:
is_finite_positive = isfinite(2.0) && 2.0 > 0
println(is_finite_positive)
println(!is_finite_positive || 2.0 == 2)

true
true


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
true
true
```

## Angles, logic, and validation before a calculation

#### A condition is a precise mathematical question

`a && b` is true only when both conditions are true; `a || b` is true when at least one is true. `xor(a,b)` is true when exactly one is true. These rules are useful for choosing mutually exclusive cases, but a model must still specify what happens at equality. A chained inequality such as $0\le\theta<90^\circ$ means both bounds must hold; we can spell it out with two comparisons joined by `&&`.

An angle needs an explicit unit. `sin` and `cos` expect radians, while `sind` and `cosd` expect degrees; `rad2deg` converts an angle from radians to degrees. The two-argument form `atan(y, x)` returns the quadrant-aware angle of the point $(x,y)$: it uses both signs, not only their ratio, so it can distinguish opposite directions. A tiny example of each is:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [28]:
println(true && false)
println(true || false)
println(xor(true, true))
println(xor(true, false))
println(sin(pi / 6))
println(sind(30.0))
println(round(rad2deg(atan(1.0, -1.0)); digits=1))

false
true
false
true
0.49999999999999994
0.5
135.0


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
false
true
false
true
0.49999999999999994
0.5
135.0
```

The point $(-1,1)$ lies at $135^\circ$, not $-45^\circ$; a one-argument `atan(y/x)` could not tell those apart. At the origin there is no physical direction; an application must state a convention. Tiny differences between equivalent floating-point calculations are not different physics.

#### Text conversion and physical validation are separate

`readline()` returns text. `parse(Float64, text)` turns valid numerical text into a number; malformed text raises an error. Successful conversion is not enough: `NaN` and `Inf` are not finite measurements, and a finite negative mass is outside our model. We validate before a square root, division, or trigonometric construction. `throw(ArgumentError("message"))` stops the calculation; merely printing an error message does not.

Before testing a physical domain, we check that the value is finite: `isfinite(value)` is `true` only for ordinary finite numbers and `false` for `NaN` or `Inf`. The following fragment converts text and validates the result as a mass:

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [29]:
text = "2.5"  # reproducible example of text supplied by readline()
mass = parse(Float64, text)   # kg
if !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite and positive"))
end
println("mass = ", mass, " kg")

mass = 2.5 kg


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
mass = 2.5 kg
```

The earlier recovery-style `tryparse`/`nothing` examples are preserved for optional input-handling reading. The core route this week uses `parse` and explicit physical domain checks; neither route requires defining a function.

## A piecewise friction model on an incline

#### Forces, signs, and assumptions

Take $+x$ uphill along a fixed incline and $+y$ normal to it. For a block of mass $m$ (kg), angle $\theta$ (radians in `sin`/`cos`), and positive gravity magnitude $g$ (m/s$^2$), gravity has downhill component $mg\sin\theta$ and normal force $N=mg\cos\theta$. Static friction can oppose the impending downhill motion up to $f_{s,\max}=\mu_sN$. If it cannot hold, kinetic friction $f_k=\mu_kN$ acts uphill and $a=(-mg\sin\theta+\mu_kN)/m$. We assume Coulomb friction, rigid contact, no drag, and $0\le\theta<90^\circ$. The model does not describe rolling, changing contact, or velocity-dependent friction. At $\theta=0$, the required static friction is zero; if $\mu_k=0$, sliding gives $a=-g\sin\theta$.

We validate before using an input. `throw(ArgumentError(...))` immediately stops the calculation and tells its caller that a documented input condition failed. It is different from `println`, which merely reports text and continues. The membership test `value isa T` asks whether a value belongs to type `T`; `Real` names Julia's family of real-number types, which includes ordinary integers and floating-point values. `Bool` needs special care here: because `true isa Real` is also `true`, a check for `Real` alone would accept `true` as a mass, so we reject it explicitly.

**Before running:** predict the value, printed text, or graph described above. This is the corresponding example from the written chapter.

In [30]:
mass = 2.0      # kg
angle = 30.0    # deg
 μ_s = 0.60
 μ_k = 0.40
g = 9.81        # m/s^2
if !(mass isa Real) || mass isa Bool || !isfinite(mass) || mass <= 0
    throw(ArgumentError("mass must be finite, positive, and non-Boolean"))
elseif !isfinite(angle) || !(0 <= angle < 90) || μ_k < 0 || μ_k > μ_s
    throw(ArgumentError("use 0 <= angle < 90 and 0 <= μ_k <= μ_s"))
end
 θ = deg2rad(angle)   # rad
normal = mass * g * cos(θ)   # N
needed_static = mass * g * sin(θ)   # N
if needed_static <= μ_s * normal
    state = "stuck"
    acceleration = 0.0   # m/s^2
else
    state = "sliding"
    acceleration = (-needed_static + μ_k * normal) / mass   # m/s^2
end
println(state, ", a = ", acceleration, " m/s^2")

stuck, a = 0.0 m/s^2


**Check:** compare the result with the explanation below. A cell displays only its final expression automatically; `println` deliberately writes text. Rerun prerequisite cells when changing an input.

Reference output in the written notes (compare with the actual cell output):

```text
stuck, a = 0.0 m/s^2
```

`stuck, a = 0.0 m/s^2`. Here $mg\sin30^\circ=9.81$ N and $\mu_sN\approx10.19$ N, so static friction is sufficient. If we change only `μ_s` to `0.40`, the result is `sliding` with negative acceleration: downhill in our declared convention. Equality belongs to the stuck branch, our explicit threshold convention.

## Recitation: a pull can reverse impending motion

We add a known uphill pull $P$ to the incline model. Before choosing a friction sign, we calculate the friction-free tendency $T=P-mg\sin\theta$. Positive $T$ means impending uphill motion; negative $T$ means impending downhill motion. Friction opposes that tendency, and static friction can hold whenever $|T|\le\mu_sN$.

We arrive able to read an `if`/`elseif`/`else` chain and convert degrees to radians. In the short guided portion we predict the direction, trace one branch, and test $T=0$. We then start the separate wall-clearance homework. The takeaway is a repeatable method: validate inputs, state a physical condition, calculate the applicable case, and check its boundary.

## Core homework: does a projectile clear a wall?

We launch from ground level with speed $v_0$, angle $\theta$, and gravity magnitude $g>0$, with $+x$ toward the wall and $+y$ upward. In the uniform-gravity, no-drag model, $$v_x=v_0\cos\theta,\quad t_w=x_w/v_x,\quad y_w=v_0\sin\theta\,t_w-\tfrac12gt_w^2,\quad v_{y,w}=v_0\sin\theta-gt_w.$$ We require $v_x>0$ before dividing. We compare $y_w$ with the wall height, and use $v_{y,w}$ to classify rising or falling. Equality in height counts as clearing; zero vertical velocity belongs to the published falling category. The computed free-flight height may be negative: for a nonnegative wall height this means failure to clear, not motion physically continuing through the ground.

We first write a prediction for one ordinary case and one boundary. Next we implement the scalar fragment, validate its documented domain, and compare the returned labels and Boolean exactly. We check heights approximately using the published tolerance. Finally we explain how the clearance decision and vertical-state decision differ: a projectile can clear while falling. The assignment sheet is the place to look for the precise input names, the tests, the wrapper convention for the returned values, and the submission instructions; neither student-defined functions nor loops are needed.

## Optional bonus: units and AC phasors

This bonus is optional, is not a prerequisite for later work, and is not examined. It is worth **one-third of a final-grade point**.

The self-contained reading develops complex numbers, magnitude, phase, conjugation, and Euler's relation before using `cis`. We choose peak amplitudes and the time convention $\exp(+i\omega t)$, set the voltage phase to zero, build $Z=|Z|\exp(i\phi_Z)$, and calculate $I=V/Z$. The measurable current is the real part of $I\exp(i\omega t)$; the complex number is a representation, not an additional measured current.

We follow the supplied Unitful examples, verify that voltage divided by resistance has current units, and check a resistor with zero phase before trying a quarter-cycle phase. We then submit the fragment assigning the complex current in amperes and run the published equivalent-unit and invalid-input checks. We explain why a positive impedance phase gives a negative current phase. This is a single-frequency linear steady-state model; switching transients and nonlinear components are outside its scope.

## Going further: total internal reflection

We use the supplied Snell-law recipe $n_1\sin\theta_1=n_2\sin\theta_2$, with positive refractive indices and angles measured from the interface normal. For $n_1>n_2$, incidence beyond $\theta_c=\arcsin(n_2/n_1)$ gives total internal reflection. We classify below-critical, critical, and above-critical examples before evaluating an inverse sine; blindly taking `asin` of a number greater than one is not a physical calculation.

We sketch repeated ideal reflections in a straight fiber and vary the launch angle. We record which choices remain guided and explain the normal used at each interface. Any supplied tracing loop is an optional preview, not Week 2 core work. Waveguide modes, dispersion, bending losses, and absorption require more physics than this ray model. This investigation is separate from the phasor bonus and is neither examined nor a prerequisite.

## Finish and reflect
Explain one result in words, name its units or assumptions where relevant, and identify a check that could reveal a mistake. The recitation, homework, optional bonus and Going further descriptions above explain the next steps. Optional background does not become an examination prerequisite.